sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuSemaphore.scala

```diff
@@ -110,29 +96,49 @@ object GpuSemaphore {
       instance = null
     }
   }
+
+  private val MAX_PERMITS = 1000
+
+  private def computeNumPermits(conf: SQLConf): Int = {
+    val concurrentStr = conf.getConfString(RapidsConf.CONCURRENT_GPU_TASKS.key, null)
+    val concurrentInt = Option(concurrentStr)
+      .map(ConfHelper.toInteger(_, RapidsConf.CONCURRENT_GPU_TASKS.key))
+      .getOrElse(RapidsConf.CONCURRENT_GPU_TASKS.defaultValue)
+    // concurrentInt <= 0 is the same as 1 (fail to the safest value)
+    // concurrentInt > MAX_PERMITS becomes the same as MAX_PERMITS
+    // (who has more than 1000 threads anyways).
+    val permits = MAX_PERMITS / math.min(math.max(concurrentInt, 1), MAX_PERMITS)
+    math.max(permits, 1)
+  }
 }
 
-private final class GpuSemaphore(tasksPerGpu: Int) extends Logging with Arm {
-  private val semaphore = new Semaphore(tasksPerGpu)
+private final class GpuSemaphore() extends Logging with Arm {
+  import GpuSemaphore._
+  private val semaphore = new Semaphore(MAX_PERMITS)
 
   // Map to track which tasks have acquired the semaphore.
-  case class TaskInfo(count: MutableInt, thread: Thread)
+  case class TaskInfo(count: MutableInt, thread: Thread, numPermits: Int)
   private val activeTasks = new ConcurrentHashMap[Long, TaskInfo]
 
   def acquireIfNecessary(context: TaskContext, waitMetric: GpuMetric): Unit = {
     withResource(new NvtxWithMetrics("Acquire GPU", NvtxColor.RED, waitMetric)) { _ =>
       val taskAttemptId = context.taskAttemptId()
       val refs = activeTasks.get(taskAttemptId)
       if (refs == null || refs.count.getValue == 0) {
-        logDebug(s"Task $taskAttemptId acquiring GPU")
-        semaphore.acquire()
+        val permits = if (refs == null) {
+          computeNumPermits(SQLConf.get)
+        } else {
+          refs.numPermits
+        }
+        logDebug(s"Task $taskAttemptId acquiring GPU with ${refs.numPermits} permits")
```

Comment:
personally a bit worried about debugging this. if I have a nsys trace and wondering why only one task running at a time when concurrent is 2 it would be hard to figure out they changed it. One option there might be to add permit number to the nvtx trace string.
That won't help the logging side if you don't have a trace.  I was trying to think of a way to just do it when it changed or once per task, but anything I've thought of seems a bit involved with having to track things.

We can just wait and see if it becomes an issue and figure out something if it does.